databuilder/backends/tpp.py

```diff
@@ -257,6 +257,101 @@ class TPPBackend(BaseBackend):
+                    WHEN mildliver IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS mildliver,
+                CASE
+                    WHEN modliv = 'YES' THEN 'Yes'
+                    WHEN modliv IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS modliv,
+                CASE
+                    WHEN chronicneu_mhyn = 'YES' THEN 'Yes'
+                    WHEN chronicneu_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS chronicneu_mhyn,
+                CASE
+                    WHEN malignantneo_mhyn = 'YES' THEN 'Yes'
+                    WHEN malignantneo_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS malignantneo_mhyn,
+                CASE
+                    WHEN chronichaemo_mhyn = 'YES' THEN 'Yes'
+                    WHEN chronichaemo_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS chronichaemo_mhyn,
+                CASE
+                    WHEN aidshiv_mhyn = 'YES' THEN 'Yes'
+                    WHEN aidshiv_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS aidshiv_mhyn,
+                CASE
+                    WHEN obesity_mhyn = 'YES' THEN 'Yes'
+                    WHEN obesity_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS obesity_mhyn,
+                CASE
+                    WHEN diabetes_type_mhyn IN ('1', '2') THEN diabetes_type_mhyn
+                    WHEN diabetes_type_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetes_type_mhyn,
+                CASE
+                    WHEN diabetescom_mhyn = 'YES' THEN 'Yes'
+                    WHEN diabetescom_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetescom_mhyn,
+                CASE
+                    WHEN diabetes_mhyn = 'YES' THEN 'Yes'
+                    WHEN diabetes_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS diabetes_mhyn,
+                CASE
+                    WHEN rheumatologic_mhyn = 'YES' THEN 'Yes'
+                    WHEN rheumatologic_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS rheumatologic_mhyn,
+                CASE
+                    WHEN dementia_mhyn = 'YES' THEN 'Yes'
+                    WHEN dementia_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS dementia_mhyn,
+                CASE
+                    WHEN malnutrition_mhyn = 'YES' THEN 'Yes'
+                    WHEN malnutrition_mhyn IN ('NO', 'NA') THEN 'No'
+                    ELSE 'Unknown'
+                END AS malnutrition_mhyn,
+                CASE
+                    WHEN smoking_mhyn IN ('Never Smoked', 'Former Smoker', 'Yes') THEN smoking_mhyn
```

Comment:
question: Do we want to be defensive here and lower-case the strings to match? Or are we reasonably assured that we'll never get anything other than the case given?

Reply:
Are you able to check by querying L3? If not, then I will 🙂 We haven't been defensive elsewhere (e.g. `Patient.DateOfBirth`, `PatientAddress.MSOACode`) so unless we have reason to, then I wouldn't be defensive here.

Reply:
> Can we just put in a test that says "we expect these columns to only take values ["X", "Y", "Z", "NA"] and if not then fail"?

Yeah, I've actually had a thought about this for categorical handling more generally. I'm going to write it up separately, but the basic idea is that the default for all CASE statements like this should be to return the original value unchanged. And then we can check while we're retrieving the data that we only ever get the expected categories.

Reply:
@evansd:

> the default for all CASE statements like this should be to return the original value unchanged

To clarify/summarise everything here and in the other thread, then that means for most of the examples in this query (NB: there are a couple of other columns that don't fit this pattern), then we have:

```
CASE
   WHEN column_name IN ('YES', 'NO', 'Unknown') THEN column_name
   WHEN column_name = 'NA' THEN 'No'
END AS column_name
```

Reply:
Or, is it actually:

```
CASE
    WHEN column_name = 'NA' THEN 'No'
    ELSE column_name
END as column_name
```

and the categorical constraint, when enforced, will validate that the data only contains the values known to us ("YES", "NO", "Unknown")?

Reply:
So, you're completely correct in your understanding of what I was suggesting. But I'm now less convinced (following discussion with BenBC) that what I was suggesting is the right approach!

I think there's a wider question we need to resolve here that it's not sensible to block your PR on. So I'd actually be inclined to stick with what you originally had and we'll  tackle the wider question separately. Sorry for all the noise!

Reply:
@evansd: I'll do a bit of refining of this PR in response to your feedback anyway before merging, because I think it's sensible.

We can still change this on a whim if we decide otherwise without breaking anything much for users, while it's still Will who's the primary consumer.